stan/math/rev/fun/mdivide_left_ldlt.hpp

```diff
@@ -13,275 +12,67 @@
-        N_(B.cols()),
-        variRefC_(reinterpret_cast<vari **>(
-            ChainableStack::instance_->memalloc_.alloc(sizeof(vari *) * B.rows()
-                                                       * B.cols()))),
-        alloc_(new mdivide_left_ldlt_alloc<R1, C1, R2, C2>()),
-        alloc_ldlt_(A.alloc_) {
-    alloc_->C_ = B;
-    alloc_ldlt_->ldlt_.solveInPlace(alloc_->C_);
-    Eigen::Map<matrix_vi>(variRefC_, M_, N_)
-        = alloc_->C_.unaryExpr([](double x) { return new vari(x, false); });
-  }
-
-  virtual void chain() {
-    matrix_d adjC = Eigen::Map<matrix_vi>(variRefC_, M_, N_).adj();
-
-    const_cast<matrix_vi &>(alloc_ldlt_->variA_).adj()
-        -= alloc_ldlt_->ldlt_.solve(adjC * alloc_->C_.transpose());
-  }
-};
-}  // namespace internal
 
 /**
  * Returns the solution of the system Ax=b given an LDLT_factor of A
  *
- * @tparam R1 number of rows in the LDLT_factor, can be Eigen::Dynamic
- * @tparam C1 number of columns in the LDLT_factor, can be Eigen::Dynamic
- *
+ * @tparam T type of B
  * @param A LDLT_factor
- * @param b Right hand side matrix or vector.
- * @return x = A^-1 b, solution of the linear system.
- * @throws std::domain_error if rows of b don't match the size of A.
+ * @param B Right hand side matrix or vector.
+ * @return x = A^-1 B, solution of the linear system.
+ * @throws std::domain_error if rows of B don't match the size of A.
  */
-template <int R1, int C1, typename EigMat,
-          require_eigen_vt<is_var, EigMat> * = nullptr>
-inline Eigen::Matrix<var, R1, EigMat::ColsAtCompileTime> mdivide_left_ldlt(
-    const LDLT_factor<var, R1, C1> &A, const EigMat &b) {
-  constexpr int R2 = EigMat::RowsAtCompileTime;
-  constexpr int C2 = EigMat::ColsAtCompileTime;
-  check_multiplicable("mdivide_left_ldlt", "A", A, "b", b);
-  if (A.cols() == 0) {
-    return {0, b.cols()};
+template <typename T1, bool alloc_in_arena, typename T2,
+          require_all_matrix_t<T1, T2>* = nullptr,
+          require_any_st_var<T1, T2>* = nullptr>
+inline auto mdivide_left_ldlt(const LDLT_factor<T1, alloc_in_arena>& A,
+                              const T2& B) {
+  using ret_val_type
+      = Eigen::Matrix<double, Eigen::Dynamic, T2::ColsAtCompileTime>;
+  using ret_type = promote_var_matrix_t<ret_val_type, T1, T2>;
+
+  check_multiplicable("mdivide_left_ldlt", "A", A.matrix().val(), "B", B);
+
+  if (A.matrix().size() == 0) {
+    return ret_type(ret_val_type(0, B.cols()));
   }
 
-  auto *baseVari
-      = new internal::mdivide_left_ldlt_vv_vari<R1, C1, R2, C2>(A, b);
+  if (!is_constant<T1>::value && !is_constant<T2>::value) {
+    arena_t<promote_scalar_t<var, T2>> arena_B = B;
+    arena_t<ret_type> res = A.ldlt().solve(arena_B.val());
 
-  Eigen::Matrix<var, R1, C2> res(b.rows(), b.cols());
-  res.vi() = Eigen::Map<matrix_vi>(baseVari->variRefC_, res.rows(), res.cols());
+    reverse_pass_callback([A, arena_B, res]() mutable {
+      promote_scalar_t<double, T2> adjB = A.ldlt().solve(res.adj());
```

Comment:
[Q] Does `.ldlt()` do something then solve happens? Like could you store that ldlt for A and then just have to do it once?

Reply:
`.ldlt()` is a member function accessing a saved LDLT decomposition, so this is what is happening.

Reply:
Avoiding multiple LDLTs is the upside of using all this `LDLT_factor` stuff.

Reply:
icic cool cool cool